### ISSUE TITLE
Artifacts don't support "file" scheme

### ISSUE THREAD

**Issue by jippi** [hashicorp/nomad#1897]
### Nomad version

Nomad v0.5.0-dev ('40d0a4e074b15ee056c5d5cc73d4a338cd19db72')

### Operating system and Environment details

OS X

### Issue

`artifact` stanza can't "download" files, example

```hcl
artifact {
  source = "/opt/deploy/www/frontend/_infrastructure/consul-template/php-fpm.ctmpl"
  destination = "local/php-fpm.ctmpl"
}
```

yields

```json
 {
     "Type": "Failed Artifact Download",
     "Time": 1477913991678771294,
     "FailsTask": false,
     "RestartReason": "",
     "SetupError": "",
     "DriverError": "",
     "ExitCode": 0,
     "Signal": 0,
     "Message": "",
     "KillTimeout": 0,
     "KillError": "",
     "KillReason": "",
     "StartDelay": 0,
     "DownloadError": "GET error: download not supported for scheme 'file'",
     "ValidationError": "",
     "DiskLimit": 0,
     "FailedSibling": "",
     "VaultError": "",
     "TaskSignalReason": "",
     "TaskSignal": ""
 }
```

**Comment by schmichael** [comment]
Nomad prevents arbitrary host file system access except for docker/rkt volume mounting and raw_exec; all of which can be disabled to lockdown a system and ensure containers can't access the host.

I'd like to keep the surface area for accessing the host file system to a minimum. Is there some compelling reason to use artifacts instead of volume mounts here?

**Comment by jippi** [comment]
@schmichael today we got all the source/config files in each project's git repo, which we zip and move to each server. So i basically want to copy the config files from the host system into the alloc dir to be used by nomad.

Using volume mounts did not work either, since template{} seem to error of missing source file before both artifact downloads or docker mount volumes even gets executed (or so it seem from my test today)

**Comment by jippi** [comment]
@schmichael though copying a file from the host system should be equally insecure as to downloading any random file from the internet, so seem like an weird omission to have :)

**Comment by dadgar** [comment]
@jippi What you can do in the meantime is running a web server hosting those files and use the artifact block. In a latter release we may add a set of directories that can be accessed. The issue right now is that if you allow arbitrary file downloads then you could access anyones alloc data which is bad.

As for downloading any file from the internet, we allow you to add the checksum for the file so you get what you expect.

**Comment by schmichael** [comment]
> copying a file from the host system should be equally insecure as to downloading any random file from the internet, so seem like an weird omission to have :)

Also worth noting preventing copying from the host is to enforce container isolation, not secure containers against malicious code (which is what the checksum feature is for).

**Comment by jippi** [comment]
@dadgar would be nice to have as an option for directly access indeed.. we run a single "monolith" environment where the entire box is trusted :) 

Also our PHP-FPM, nginx and redis configuration is contained in the repo we deploy through nomad, so those files will only be available either directly on the host disk or inside the docker container, which the current Nomad setup doesn't work with out of the box. 

I'm not sure about the best practice on this, but not being able to use the docker image content or the host file system seem very limiting to me. 

I completely get this is the first release with this (awesome!) feature, and it need time to mature and grow into a more feature rich and secure solution over time :)

**Comment by jippi** [comment]
Additionally, having nomad in production / staging and development require different level of strictness.. for example running nomad on OS X to mirror production, running a separate server to serve files from the local environment is ... not ideal either.. it should be as simple as `nomad run <file>` and done.

Having it as a flag similar to raw_exec would be nice :)

**Comment by jippi** [comment]
@dadgar @schmichael 

What I'm looking for is to do one of the following things, to keep moving parts to a minimum

- be able to artifact copy from the host fs
- be able to artifact copy from a docker volume

all our app code lives as directories on the docker host machines, and we use a volume to map the code into our "execution container"

our repos contain all nginx/php-fpm/php-cli configuration files for that project in particular, so I need to find a way to get the configuration from a project directory into the alloc directory, without having to spin up an local nginx server *just* to work around the limitation of copy being disallowed. 

i think it's a pretty tall requirement to have to service template files from a webserver, when the files already exist locally both in a docker volume and in the host filesystem. 

any advice on how to make this a more sane setup is highly welcome

**Comment by dadgar** [comment]
Hey,

We will get the artifact from host working in 0.5.1. It is just a bit too much to plumb for 0.5

**Comment by jippi** [comment]
@dadgar <3 ! :)

**Comment by mongrelion** [comment]
@dadgar I'm running `0.5.1` here and I still get this error. Did it land and I'm missing something? :)

**Comment by dradtke** [comment]
I just got a `GET error: download not supported for scheme 'file'` on `0.5.2` as well.

**Comment by schmichael** [comment]
Sadly I didn't get to it in time for 0.5.3. I'll try again for 0.5.4.

We want to implement it such that the client has a whitelist of paths artifacts can be copied from to maintain a tunable level of encapsulation between tasks.

**Comment by jippi** [comment]
did it get in yet @schmichael ? :)

**Comment by danlsgiga** [comment]
I'm also looking forward for this feature. Either being able to copy files from the host to the task or being able to provide an additional list of files to be copied to the chroot in the task level instead of the client config.

**Comment by mongrelion** [comment]
I have a use case where one would have some templates already on the host and they just need to make it to the allocation directory.
A workaround for sure is the web server to serve these files but sounds like an overkill for this case. You would be introducing yet another service into the moving parts of your system.

**Comment by tkoestler** [comment]
Is this ever going to be addressed?

**Comment by danapoklepovich** [comment]
Is using the local filesystem as the artifact source on the horizon at all? I need nomad to work with the artifacts already on the servers box as contacting an external api isn't possible and spinning up a web server does seem like over kill especially when the go getter library does support the file scheme. Or is there some specific reason for this option being excluded? I'm running 0.8.6 and still getting this error message.

**Comment by LordFPL** [comment]
Any news on that ?

**Comment by andye2004** [comment]
Hello, is there anyone there? Would really like to know if this is ever likely to be implemented? If not could someone update the issue with appropriate comments and then close it off?

**Comment by cgbaker** [comment]
@danapoklepovich : as @schmichael pointed out, we're hesitant to add this feature because it allows for arbitrary content from the client (including, other tasks) to be copied into a task. Towards @andye2004's comment, simply enabling the `file` schema without a safelist allows **anyone job submitter** access to the potentially the entire client filesystem; contrast this with configuring the chroot_env, which is performed during the privileged act of running the client.

I'll ask the dev team about a timetable for adding this capability.

**Comment by cgbaker** [comment]
Additionally, we are designing a number of features around volume support, including a host volume capability that should address these use cases without compromising security.

**Comment by dantoml** [comment]
#5377 is an issue you can follow if you're interested in support for Host Volume mounts 😄

**Comment by andye2004** [comment]
@cgbaker @dantoml Thanks guys, appreciate the follow up. Volumes sounds like a much better solution IMO. I'll keep an eye on #5377, got excited when I saw the ticket and then a bit disappointed to see it has only just been opened.....

**Comment by preetapan** [comment]
Closing this, Host volumes is now available in Nomad 0.10 and we don't intend to support the `file` schema for the reasons outlined above.

**Comment by johnzhanghua** [comment]
@preetapan  @endocrimes  Host volume is not going to help this issue.  For detail check 
https://github.com/hashicorp/nomad/issues/6846

**Comment by schmichael** [comment]
I'll reopen this as #6846 outlines a use case host volumes do not cover.

That being said: even if the security issues are solved (perhaps with an allow list), the most naive approach's user experience is not ideal:

Unlike host volumes, the scheduler has no knowledge of what files exist on a client's filesystem, so there's no ability to properly schedule a job onto a client that has the files the job expects to exist. This means a user would have to put the files on every node or remember to use node metadata and job constraints when referencing files on the host filesystem.

This isn't a blocker per se, but I'm still unsure if the use cases solved by the `file` scheme aren't better suited in other ways (eg "build" support for baking artifacts into images, better image management support to avoid the need for locally sourced images, etc).

**Comment by johnzhanghua** [comment]
@schmichael  Thank you. 

> This means a user would have to put the files on every node or remember to use node metadata and job constraints when referencing files on the host filesystem.

That can be the use cases, or using GlusterFs. 

> This isn't a blocker per se, but I'm still unsure if the use cases solved by the file scheme aren't better suited in other ways (eg "build" support for baking artifacts into images, better image management support to avoid the need for locally sourced images, etc).

To load the images from local is a valid requirement when the images are within the cluster nodes.  From "artifact" perspective, it looks focus loading outside of the cluster nodes.

**Comment by salotz** [comment]
Seems like minimally this would be useful for troubleshooting locally, even if it is insecure for production.

**Comment by NarimanAB** [comment]
@schmichael  "I'd like to keep the surface area for accessing the host file system to a minimum."
hm... to access files on the host is bad but to access them on an arbitrary url is good? I dont see the point

**Comment by schmichael** [comment]
> I think this feature would be nice to have for ease of local development. nomad agent -dev and local artifact sources would make it super easy to get up and running, and would be useful for experimentation as well as rapid iteration.
-- @dradtke 

We could add an agent configuration parameter to enable `file` artifact access, enable that for `-dev` and disable it by default otherwise.

However, this introduces even more skew than already exists between `-dev` and "real" Nomad agents. In particular in introduces a difference that means your local job file is unlikely to work _anywhere_ else, thus limiting the usefulness of your local testing.

To make matters worse we make absolute artifact paths relative to the task directory implicitly, so it would be difficult for accidental deployments of dev jobs to fail anywhere other than at runtime on clients.

Getting the optimal UX here seems extremely hard given the corner we've backed ourselves into with artifact path handling.

> hm... to access files on the host is bad but to access them on an arbitrary url is good? I dont see the point
-- @NarimanAB 

Let me restate your question to ensure I'm understanding it:

> Why is accessing a local file more dangerous than accessing a remote file?

For the application: it's not! Those are more or less equivalent from a threat model perspective.

However what I'm worried about here is not the application's safety, but the *host's and the cluster's.* Applications are *untrusted* and the core value proposition of containers is that it lets you run applications with low-to-zero trust in them.

We can't trust that apps won't do "bad things" with full host filesystem access. They could read Nomad's mTLS certificates, impersonate a Nomad client node, accept allocations, and steal Vault tokens for those allocations. They could read the task directories of other tasks and steal their Vault tokens.

Even if you trust your app developers, a core value proposition of containers is that _even if the apps are taken over and an attacker has remote code execution the scope of the threat is limited to the capabilities of that container._ Host filesystem access allows numerous opportunities for attackers to elevate their attack outside the scope of the vulnerable app.

That being said there's a reason this issue is still open! Host filesystem access is not an all-or-nothing proposition!

Host volumes and CSI are 2 ways to give apps limited file access and restrict their capabilities to some cluster operator defined scope.

We had hoped that would be enough, but as pointed out in #6846 there's still a siginficant use case we didn't cover: Container/vm images.

This issue covering the `file` artifact is one possible way we could solve that, but there's enough downsides in safely implementing that that it hasn't been a priority yet. If container/vm images are the primary use case, perhaps there's a first class solution for that specific use case that is easier to secure than the `file` artifact scheme.

**Comment by msuarezd** [comment]
Hi @schmichael,
I totally understand and agree with your explanation: applications should not be trusted and they should not be able to access the host's filesystem. However (in this case) it is about an artifact stanza in the nomad job. Which means the application itself would not be able to access the hosts filesystem anyway. Only somebody with client access and rights to deploy. Am I right or do I oversee something?

From my perspective that would be extreme helpful for testing. I think of this is a feature which could be disabled by default, but can be explicitely enabled, analogous to the raw_exec driver. Would that be a good compromise?

**Comment by schmichael** [comment]
> However (in this case) it is about an artifact stanza in the nomad job. Which means the application itself would not be able to access the hosts filesystem anyway. Only somebody with client access and rights to deploy. Am I right or do I oversee something?

Great point @msuarezd ! You're absolutely right!

> I think this is a feature which could be disabled by default, but can be explicitely enabled like you proposed above. Analogous to the raw_exec driver. That would be a good compromise.

I'm not against this per se, but every design I come up with mentally ends up looking like host volumes. Unrestricted `file` access, even disabled by default, seems problematic as it allows a job operator access to things like the node's TLS certificates and all allocation's vault tokens. Is there a reason not to use host volumes other than issue #6846 where they're not available early enough for use in fields like Docker's [`load`](https://www.nomadproject.io/docs/drivers/docker#load)?

We could _only_ enable it with the `-dev` flag, but that's going to be a frustrating thing for people when they run into it.

**Comment by msuarezd** [comment]
Sorry for the late reply @schmichael . 
I must confess I did not try host volumes for this, but it looks like usable workaround for me. Thanks for pointing it out!

**Comment by manibk31** [comment]
I've been following this thread for a while now, my requirements are to be able to load a docker tar generated using docker save command and if the file scheme is not supported, I understand the there is a need to arrange for the images to be downloaded via a URI.

Considering the fact that host volumes have been proposed as one of the solutions for this problem somewhere above, my question is how to go about it? If a host volume is specified  in the job description and an absolute path pointing to the tar bundle in the host volume is passed as an artifact source, would the scheme in this case be accepted exceptionally?

**Comment by natto1784** [comment]
I'm facing exactly the same problem as manlbk31

**Comment by joeyguerra** [comment]
Does host volume even work with raw_exec?

**Comment by exFalso** [comment]
Ditto, same issue with loading docker tars. I'm not even sure how we're supposed to use the `load` docker config argument, seeing as it reads the tar from the `local/` directory... How can we workaround this?

**Comment by exFalso** [comment]
In fact, I'd propose changing the documentation of `load` from saying

> [load](https://www.nomadproject.io/docs/drivers/docker#load) - (Optional) Load an image from a `tar` archive file instead of from a remote repository. Equivalent to the `docker load -i <filename>` command. If you're using an `artifact` block to fetch the archive file, you'll need to ensure that Nomad keeps the archive intact after download.

to

> [load](https://www.nomadproject.io/docs/drivers/docker#load) - (Optional) Load an image from a `tar` archive file instead of from a remote repository. Equivalent to the `docker load -i <filename>` command. **Warning!** This option causes Nomad to load images from the `local/` allocation directory even if you specify an absolute path, so the only way to use this option is in tandem with the `artifact` stanza. And because we disabled local file access in the `artifact` stanza for security reasons, you can only use this option if you are downloading the tar from a remote location.

**Comment by otto-dev** [comment]
The artifact documentation refers me to the go-getter docs, [which says local files are supported](https://github.com/hashicorp/go-getter#url-format). But then we get `download not supported for scheme 'file'`.

**Comment by krabiswabbie** [comment]
Easy things like using a binary assets from a private `github` repository becomes extremely complicated. First there is no way to download an asset by the direct link. The workaround here is to use third-party tools like `gruntwork-io/fetch` within Init task to have it downloaded. And even with that I've just discovered that there is no way to get things done even if `go-getter` manual clearly says it's possible! That's a bummer..

**Comment by maxadamo** [comment]
> The artifact documentation refers me to the go-getter docs, [which says local files are supported](https://github.com/hashicorp/go-getter#url-format). But then we get `download not supported for scheme 'file'`.

I quote this. I have spent lot of time trying and re-trying, because go-getter supports local files, and Nomad documentation doesn't say that it's not supported. 

Would it make sense and is it feasible to allow interpolation of templates, for keys downloaded from Consul or from Nomad? 
For instance there could be a parameter, like as `template = true/false`

**Comment by otto-dev** [comment]
Leaving a little workaround here: In some cases you might be able to work around this issue by using the `file` function inside a `template` block to write a file to a job.  https://github.com/hashicorp/consul-template/blob/main/docs/templating-language.md#file

You might have to change some security setting in the nomad config for this - I'm going off memory.

**Comment by tgross** [comment]
> I quote this. I have spent lot of time trying and re-trying, because go-getter supports local files, and Nomad documentation doesn't say that it's not supported.

The docs say:
> Nomad supports downloading http, https, git, hg and S3 artifacts.

Maybe we could make that a larger/more prominent warning though if more than one person has wasted their time realizing that

**Comment by maxadamo** [comment]
> > I quote this. I have spent lot of time trying and re-trying, because go-getter supports local files, and Nomad documentation doesn't say that it's not supported.
> 
> The docs say:
> 
> > Nomad supports downloading http, https, git, hg and S3 artifacts.
> 
> Maybe we could make that a larger/more prominent warning though if more than one person has wasted their time realizing that

in this case it's my bad. I didn't read carefully. 
At same time I've set this parameter: 
```json
  "client": {
    "template": {
      "disable_file_sandbox": false
    }
    ...
  }
```

I try to use this in the template block of my job:

```hcl
      template {
        data          = <<EOF
{{ file "/depot/nomad_shared/templates/prod/prometheus/prometheus.yml.tpl" }}
        EOF
        destination   = "local/prometheus.yml"
        change_mode   = "signal"
        change_signal = "SIGHUP"
      }
```

I reload nomad daemon. 
I try to run my job and I get this error: 

```
2023-08-04T15:14:26+02:00  Killing          Template failed: (dynamic): execute: template: :1:3: executing "" at <file "/depot/nomad_shared/templates/prod/prometheus/prometheus.yml.tpl">: error calling file: '/depot/nomad_shared/templates/prod/prometheus/prometheus.yml.tpl' is outside of sandbox
```

**Comment by tgross** [comment]
@maxadamo the error message is saying the file is outside the template sandbox, and you have `disable_file_sandbox: false` which means the sandbox is active. But now we're getting into issues with `template` rather than `artifact`, so if you're still having trouble after adjusting that field, let's open a new issue for it.

**Comment by maxadamo** [comment]
@tgross I deleted my message before you replied 😺  I flipped true and false. Thanks a lot for your reply.

**Comment by jblachly** [comment]
> > I quote this. I have spent lot of time trying and re-trying, because go-getter supports local files, and Nomad documentation doesn't say that it's not supported.
> 
> The docs say:
> 
> > Nomad supports downloading http, https, git, hg and S3 artifacts.
> 
> Maybe we could make that a larger/more prominent warning though if more than one person has wasted their time realizing that

I'm here from the future to indicate that this is still a problem, not only functionally but with the documentation as well.

My use case is with the "exec" driver -- I have a binary that does not live in the `/usr` tree (or any other chroot'ed filesystem) but that I wish to copy into the `local/` of the task so I can run it.

**Comment by jippi** [comment]
Closing since I don't use Nomad, and the issue is almost 10 years old 🤷

**Comment by jblachly** [comment]
Why not leave it open since many of us are still waiting on this?

**Comment by jippi** [comment]
Because GitHub don't allow me to hide it from my dashboard without closing it - and after 9 years, its unlikely to happen considering how simple the thing is to implement - it would have been done by now.

**Comment by jguerractr-hue** [comment]
ah man.
